### PR TITLE
contains exactly in any order

### PR DESCRIPTION
**Pull request by lpandzic**

### CODE REVIEW

File: src/main/java/org/assertj/core/api/ObjectEnumerableAssert.java [hunk 1]
```diff
@@ -127,6 +127,30 @@
...
+   * Verifies that the actual group contains exactly the given values and nothing else, <b>in any order</b>.<br>
+   *
+   *
+   * <p>
+   * Example :
+   * <pre><code class='java'> // an Iterable is used in the example but it would also work with an array
+   * Iterable&lt;Ring&gt; elvesRings = newArrayList(vilya, nenya, narya, vilya);
+   *
+   * // assertion will pass
+   * assertThat(elvesRings).containsExactly(vilya, vilya, nenya, narya);
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/ObjectEnumerableAssert.java]: `containsExactlyInAnyOrder` ? ;-)

File: src/main/java/org/assertj/core/error/ShouldContainExactlyInAnyOrder.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+          "to contain exactly in any order:%n" +
+          "  <%s>%n" + (errorType == NOT_FOUND ? 
+          "but could not find the following elements:%n" : "but the following elements were unexpected:%n") +
+          "  <%s>%n%s",
+          actual, expected, notFoundOrNotExpected, comparisonStrategy);
+    // @format:on
+  }
+
+  public enum ErrorType {
+    NOT_FOUND, NOT_EXPECTED
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/error/ShouldContainExactlyInAnyOrder.java]: I would rename these to NOT_FOUND_ONLY, NOT_EXPECTED_ONLY

File: src/test/java/org/assertj/core/internal/iterables/Iterables_assertContainsExactlyInAnyOrder_Test.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+  @Test
+  public void should_pass_if_actual_contains_exactly_given_values() {
+    iterables.assertContainsExactlyInAnyOrder(someInfo(), actual, array("Luke", "Yoda", "Leia"));
+  }
+
+  @Test
+  public void should_pass_if_actual_contains_given_values_exactly_with_null_elements() {
+    iterables.assertContainsExactlyInAnyOrder(someInfo(), actual, array("Luke", "Yoda", "Leia"));
+    actual.add(null);
+    iterables.assertContainsExactlyInAnyOrder(someInfo(), actual, array("Luke", "Yoda", "Leia", null));
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/internal/iterables/Iterables_assertContainsExactlyInAnyOrder_Test.java]: I would just shuffle the order

**Comment by joel-costigliola** [comment]
Good work ! It only misses ObjectArrays_assertContainsExactlyInAnyOrder_Test if I'm not mistaken.

File: src/main/java/org/assertj/core/api/ObjectEnumerableAssert.java (hunk 1 above)
Comment (lpandzic) [inline src/main/java/org/assertj/core/api/ObjectEnumerableAssert.java]: :+1:

File: src/main/java/org/assertj/core/error/ShouldContainExactlyInAnyOrder.java (hunk 2 above)
Comment (lpandzic) [inline src/main/java/org/assertj/core/error/ShouldContainExactlyInAnyOrder.java]: It's the same as in ShouldContainOnly#ErrorType, should I rename there too?

File: src/test/java/org/assertj/core/internal/iterables/Iterables_assertContainsExactlyInAnyOrder_Test.java (hunk 3 above)
Comment (lpandzic) [inline src/test/java/org/assertj/core/internal/iterables/Iterables_assertContainsExactlyInAnyOrder_Test.java]: :+1:

File: src/main/java/org/assertj/core/error/ShouldContainExactlyInAnyOrder.java (hunk 2 above)
Comment (lpandzic) [inline src/main/java/org/assertj/core/error/ShouldContainExactlyInAnyOrder.java]: done

**Comment by lpandzic** [comment]
> Good work ! It only misses org.assertj.core.internal.objectarrays.ObjectArrays_assertContainsExactlyInAnyOrder_Test if I'm not mistaken.

Added. I would add `containsExactlyInAnyOrder` to 'AbstractMapAssert' also but those Array test duplications killed my enthusiasm (and time) :/

**Comment by joel-costigliola** [comment]
I so wish primitive types were Objects - I'm probably not the only one when you see all the work needed for primitive type stream in the Stream API.

**Comment by lpandzic** [comment]
Java 10 and Valhalla... Soon™.

**Comment by joel-costigliola** [comment]
Integrated in 2.x (nothing really java 8 specific in the assertion).
thanks !

**Pull request closed without merging**